File: generated_code/cpp/solution_1917_3.cpp
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
        // Lambda to calculate the potential gain in pass ratio if we add 1 student to a class
        auto getGain = [](double pass, double total) {
            return ((pass + 1.0) / (total + 1.0)) - (pass / total);
        };

        // Max-heap stores elements as {gain, {pass, total}}
        // We sort by the gain in descending order to always pick the class with the maximum impact
        priority_queue<pair<double, pair<int, int>>> maxHeap;

        // Populate the heap with the initial classes and their potential gains
        for (const auto& c : classes) {
            double gain = getGain(c[0], c[1]);
            maxHeap.push({gain, {c[0], c[1]}});
        }

        // Greedy approach: assign each extra student one by one to the class yielding the highest gain
        while (extraStudents > 0) {
            auto [currentGain, studentCounts] = maxHeap.top();
            maxHeap.pop();

            int pass = studentCounts.first;
            int total = studentCounts.second;

            // Increment the student counts for the chosen class
            pass++;
            total++;
            extraStudents--;

            // Recalculate the next potential gain and push back into the heap
            maxHeap.push({getGain(pass, total), {pass, total}});
        }

        // Calculate the final total pass ratio across all classes
        double totalPassRatio = 0.0;
        while (!maxHeap.empty()) {
            auto [_, studentCounts] = maxHeap.top();
            maxHeap.pop();
            totalPassRatio += (double)studentCounts.first / studentCounts.second;
        }

        // Return the average pass ratio
        return totalPassRatio / classes.size();
    }
};

File: generated_code/python/solution_1917_3.py
import heapq
from typing import List

class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        """
        To maximize the average pass ratio, we need to maximize the sum of pass ratios.
        This is a greedy problem where we want to allocate each extra student to the class
        that yields the highest *incremental gain* in its pass ratio.
        
        The incremental gain for a class with (pass, total) is:
        gain = ((pass + 1) / (total + 1)) - (pass / total)
        
        We can use a max-heap to efficiently fetch and update the class with the highest gain.
        Since Python's heapq is a min-heap, we push negative gains.
        
        Time Complexity: O(N log N + K log N), where N = len(classes) and K = extraStudents.
        Space Complexity: O(N) to store the elements in the heap.
        """
        
        # Helper function to calculate the gain of adding one passing student
        def get_gain(p: int, t: int) -> float:
            return (p + 1) / (t + 1) - p / t

        # Initialize the max-heap with (-gain, pass, total)
        # We store negative gain because heapq is a min-heap by default
        max_heap = []
        for p, t in classes:
            gain = get_gain(p, t)
            max_heap.append((-gain, p, t))
            
        heapq.heapify(max_heap)

        # Greedily allocate each extra student to the class with the maximum gain
        for _ in range(extraStudents):
            neg_gain, p, t = heapq.heappop(max_heap)
            p += 1
            t += 1
            # Recalculate the gain for the updated class and push it back
            new_gain = get_gain(p, t)
            heapq.heappush(max_heap, (-new_gain, p, t))

        # Calculate the final average pass ratio
        total_ratio_sum = 0.0
        for _, p, t in max_heap:
            total_ratio_sum += p / t
            
        return total_ratio_sum / len(classes)